junit-platform-suite-engine/src/main/java/org/junit/platform/suite/engine/SuiteLauncherDiscoveryRequestBuilder.java

```diff
@@ -0,0 +1,182 @@
+import static org.junit.platform.commons.support.AnnotationSupport.findAnnotation;
+import static org.junit.platform.commons.support.AnnotationSupport.findRepeatableAnnotations;
+import static org.junit.platform.engine.discovery.ClassNameFilter.excludeClassNamePatterns;
+import static org.junit.platform.engine.discovery.ClassNameFilter.includeClassNamePatterns;
+import static org.junit.platform.engine.discovery.DiscoverySelectors.selectClasspathResource;
+import static org.junit.platform.engine.discovery.DiscoverySelectors.selectDirectory;
+import static org.junit.platform.engine.discovery.DiscoverySelectors.selectFile;
+import static org.junit.platform.engine.discovery.PackageNameFilter.excludePackageNames;
+import static org.junit.platform.engine.discovery.PackageNameFilter.includePackageNames;
+import static org.junit.platform.launcher.EngineFilter.excludeEngines;
+import static org.junit.platform.launcher.EngineFilter.includeEngines;
+import static org.junit.platform.launcher.TagFilter.excludeTags;
+import static org.junit.platform.launcher.TagFilter.includeTags;
+
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import org.junit.platform.engine.UniqueId;
+import org.junit.platform.engine.discovery.ClassSelector;
+import org.junit.platform.engine.discovery.ClasspathResourceSelector;
+import org.junit.platform.engine.discovery.ClasspathRootSelector;
+import org.junit.platform.engine.discovery.DiscoverySelectors;
+import org.junit.platform.engine.discovery.FilePosition;
+import org.junit.platform.engine.discovery.FileSelector;
+import org.junit.platform.engine.discovery.ModuleSelector;
+import org.junit.platform.engine.discovery.PackageSelector;
+import org.junit.platform.engine.discovery.UriSelector;
+import org.junit.platform.launcher.LauncherDiscoveryRequest;
+import org.junit.platform.launcher.core.LauncherDiscoveryRequestBuilder;
+import org.junit.platform.suite.api.Configuration;
+import org.junit.platform.suite.api.ExcludeClassNamePatterns;
+import org.junit.platform.suite.api.ExcludeEngines;
+import org.junit.platform.suite.api.ExcludePackages;
+import org.junit.platform.suite.api.ExcludeTags;
+import org.junit.platform.suite.api.IncludeClassNamePatterns;
+import org.junit.platform.suite.api.IncludeEngines;
+import org.junit.platform.suite.api.IncludePackages;
+import org.junit.platform.suite.api.IncludeTags;
+import org.junit.platform.suite.api.SelectClasses;
+import org.junit.platform.suite.api.SelectClasspathResource;
+import org.junit.platform.suite.api.SelectClasspathRoots;
+import org.junit.platform.suite.api.SelectDirectories;
+import org.junit.platform.suite.api.SelectFile;
+import org.junit.platform.suite.api.SelectModules;
+import org.junit.platform.suite.api.SelectPackages;
+import org.junit.platform.suite.api.SelectUris;
+
+final class SuiteLauncherDiscoveryRequestBuilder {
```

Comment:
Can we share this with `junit-platform-runner`?

Reply:
While the name might create the impression that it can be used by a user to build requests to discover test suites, it is only a utility that the suite engine uses to create launcher discovery requests to discover tests the tests in a suite. In essence it maps suite annotations to a launcher discovery request.

Perhaps a better name is in order.

Reply:
This is exactly what `junit-platform-runner` does with the annotations, isn't it?

Reply:
Ah I see. It'd be a single class/module. I'll see about extracting that after everything else.

Reply:
On second thought. This would require solving quite a few problems that I'd consider out of scope of this PR. I'm also not convinced the effort to remove a relatively trivial amount of not-quite-duplication is worth the added complexity of adding another module and satisfying two somewhat dissimilar use-cases..

While functionally almost the same the JUnitPlatformRunner is applying some introspection to work out if it is used as a suite or not. Unfortunately this is determined by the implicit use of either `@SelectPackages` or `@SelectClasses` rather then an explicit annotation (e.g. `@Suite`).

```java
	private LauncherDiscoveryRequest createDiscoveryRequest() {
		List<DiscoverySelector> selectors = getSelectorsFromAnnotations();

		// Allows to simply add @RunWith(JUnitPlatform.class) to any test case
		boolean isSuite = !selectors.isEmpty();
		if (!isSuite) {
			selectors.add(selectClass(this.testClass));
		}

		LauncherDiscoveryRequestBuilder requestBuilder = request().selectors(selectors);
		addFiltersFromAnnotations(requestBuilder, isSuite);
		return requestBuilder.build();
	}
```	

This could be worked around by making the `SuiteLauncherDiscoveryRequestBuilder` build a `LauncherDiscoveryRequestBuilder` rather then a `DiscoveryRequestBuilder`. Using that `LauncherDiscoveryRequestBuilder` to build the `LauncherDiscoveryRequest`. Inspecting the  `LauncherDiscoveryRequest` and adding the defaults when absent. 

This would also require making `IncludeClassNameFilter` and friends public.

```java
                LauncherDiscoveryRequestBuilder requestBuilder = suiteLauncherDiscoveryRequestBuilder.build();
                LauncherDiscoveryRequest build = requestBuilder.build();
		List<ClassSelector> classSelectors = build.getSelectorsByType(ClassSelector.class);
		List<PackageSelector> packageSelectors = build.getSelectorsByType(PackageSelector.class);
		List<IncludeClassNameFilter> includeClassNameFilters = build.getFiltersByType(IncludeClassNameFilter.class);

		// Allows to simply add @RunWith(JUnitPlatform.class) to any test case
		boolean isSuite = !classSelectors.isEmpty() && !packageSelectors.isEmpty();
		if (!isSuite) {
			requestBuilder.selectors(selectClass(this.testClass));
		} else {
			if(includeClassNameFilters.isEmpty()) {
				requestBuilder.filters(includeClassNamePatterns(STANDARD_INCLUDE_PATTERN_ARRAY));
			}
		}
		return requestBuilder.build();
```		

From the other side, the `SuiteLauncherDiscoveryRequestBuilder` currently supports adding a requests from a class by looking at it's annotations and adding a request from a unique id. The latter is needed because suites may be selected by their unique id (the class) or by the unique id of a test in that suite, or both.

This could be worked around by making the `SuiteLauncherDiscoveryRequestBuilder` build a `LauncherDiscoveryRequestBuilder` rather then a `DiscoveryRequestBuilder` but would introduce another layer of indirection as the `SuiteTestDescriptor` would have store track unique ids and test classes and build the discovery request after all selectors have been resolved delaying feedback on incorrectly used annotations.

Reply:
Ooh. I found a solution!

Reply:
Done. See: `SuiteLauncherDiscoveryRequestBuilder`.